ES6 Updates

lucasvallenet opened the pull request:
Clean-up utility functions:

- Use ES6 syntax
- Update some functions to use ES6
- Add block comments to every functions

I did not edit the `visibility.js` to avoid conflicts with [PR#99](https://github.com/locomotivemtl/locomotive-boilerplate/pull/99)

mcaskill commented on assets/scripts/utils/is.js:
```diff
@@ -1,37 +1,67 @@
 
-    return a === b;
-}
+const isArrayLike = x => arrayLikePattern.test(toString.call(x))
```
This expands the scope of this function beyond `Array` and `FileList` by checking for `length` which is the primary feature of "array-like" values.

```suggestion
const isArrayLike = x => (x && typeof x !== 'function' && isArrayLikeLength(x.length))

/**
 * Checks if argument is a valid array-like length.
 *
 * @param   {*} value - The value to check.
 * @returns {boolean}
 */
const isArrayLikeLength = x => (typeof x === 'number' && x > -1 && x % 1 == 0)
```

mcaskill commented on assets/scripts/utils/maths.js:
```diff
@@ -1,3 +1,39 @@
+const invlerp = (x, y, a) => clamp((v - x)/(a - x))
+
+
+// Export
```
```suggestion
```

mcaskill commented on assets/scripts/utils/html.js:
```diff
@@ -1,120 +1,144 @@
+ * @return {array}          parent nodes
+ */
+
+const queryClosestParent = ($el, selector) => {
```
Either remove this function in favour of `Element.closest()` or have it be an alias of it.

I prefer we remove this custom function.

mcaskill replied:
@lucasvallenet 👋

mcaskill commented on assets/scripts/utils/array.js:
```diff
@@ -1,101 +1,124 @@
+ * @return {array}  array converted
+ */
+
+const toArray = arrayLike => {
```
This should be replaced with `Array.from()`.

lucasvallenet replied:
If it's to return a function that only does `array => Array.from(array)` we can definitely delete this one

mcaskill commented on assets/scripts/utils/array.js:
```diff
@@ -1,101 +1,124 @@
-export function removeFromArray ( array, member ) {
-    if ( !array ) {
-        return;
+const removeFromArray = (array, value) => {
```
Should this be replaced with `Array.prototype.filter()`?

```js
items.filter(item => item !== itemToRemove)
```

lucasvallenet replied:
The filter function doesn't mutate the array, it only return the array filtered

I'm not really sure if someone still uses these helpers (`array.js` / `transform.js` / `is.js`), we can probably delete some unsued functions. ES6 comes in handy for some use case such as this one.

mcaskill commented on assets/scripts/utils/array.js:
```diff
@@ -1,101 +1,124 @@
 
-export function arrayContentsMatch ( a, b ) {
-    let i;
+const arrayContains = (array, value) => array.find(item => item === value)
```
This should be replaced with `Array.prototype.includes()`.

lucasvallenet replied:
Probably to delete too

mcaskill commented on assets/scripts/utils/array.js:
```diff
@@ -1,101 +1,124 @@
+ * @return {void}
+ */
+
+const addToArray = (array, value) => {
+    if (array.indexOf(value) === -1) {
+        array.push(value)
     }
 }
```
Should this be replaced with `Set`?

```js
const set = new Set();

set.add(1) // Set [ 1 ]
set.add(5) // Set [ 1, 5 ]
set.add(5)

Array.from(set)
```

I [searched](https://bitbucket.org/search?account=%7B3c157d8e-1c3f-4d33-bbaf-0d83eb71672b%7D&q=addToArray%20ext%3Ajs%20-path%3A%2Fwww%20-path%3Autils) our client projects and found only one occurrence of this function.

mcaskill commented on assets/scripts/utils/array.js:
```diff
@@ -1,101 +1,124 @@
-import { isArray } from './is';
+import { isArray } from './is'
```
This should be replaced with `Array.isArray()`.

lucasvallenet replied:
Import to delete, it is not used within the file.
The `isArray` function probably too as we have the ES6 syntax

mcaskill reviewed (commented):
We should implement ESLint to format the codebase into something consistent.

lucasvallenet commented:
+1 for adding an ES linter warning on save/commit

lucasvallenet commented:
I completely deleted the `array.js` util.
They are almost never used and a single ES6 line is possible as an alternative :

```
addToArray: new Set(array.push(value))
arrayContains: array.includes(value)
lastItemOfArray: array[array.length - 1]
removeFromArray: array.filter(item => item !== value)
toArray: Array.from(array)
findByKeyValue: array.find(item => item[key] === value)
cloneArray: [...array]
shuffleArray: array.sort(() => 0.5 - Math.random())
```

mcaskill commented on assets/scripts/utils/maths.js:
```diff
@@ -31,9 +31,22 @@ const lerp = (x, y, a) => x * (1 - a) + y * a
 
 
-// Export
+/**
+ * Round number to n decimal
+ * @param {number}  x   - float number to round
+ * @param {number}  n   - integer number of decimals
+ * @return {number} rounded number
+ */
+
+const roundNumber = (x, n = 2) => {
+    const decimals = 10 ** n
+    return Math.round(x * decimals)/decimals
+}
```
I suggest using [`Number.prototype.toPrecision()`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Number/toPrecision) instead.

```suggestion
```

mcaskill replied:
@lucasvallenet 👋

arnvvd replied:
I suggest keeping this function because it returns a number, and [Number.prototype.toPrecision()](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Number/toPrecision) returns a string 🤷‍♂️

mcaskill replied:
Oh! I did not know that.

In that case, what about this:

```suggestion
/**
 * Round number to the specified precision.
 *
 * This function is necessary because `Number.prototype.toPrecision()`
 * and `Number.prototype.toFixed()`
 *
 * @param  {number}  number      - The floating point number to round.
 * @param  {number}  [precision] - The number of digits to appear after the decimal point.
 * @return {number} The rounded number.
 */
const roundNumber = (number, precision = 2) => {
    return Number.parseFloat(number.toPrecision(precision));
}
```

mcaskill reviewed (changes requested):
Note that a few suggestions from my initial review are still active but are hidden by GitHub.

lucasvallenet commented:
- Added `config.js` with constant values
- Added a throttle function and updated the filename `debounce.js` to `tickers.js`
- Added `events.js` with `addStartEvent` and `andEndEvents` utils to easily bind an even on start/end with a debounce delay. Also added `addScrollUpEvent` and `addScrollDownEvent` to easily listen when the scroll direction updates

### Usage of events :

**addStartEvent**
```
import { addStartEvent } from './utils/events'

// Detect document mousemove start with 1000ms delay
addStartEvent(document, 'mousemove', 1000)

document.addEventListener('mousemoveStart', () => {
   // ...callback
})
```

**addEndEvent**
```
import { addEndEvent } from './utils/events'

// Detect window resize end with 200ms delay (default value)
addEndEvent(window, 'resize')

window.addEventListener('resizeEnd', () => {
   // ...callback
})
```

**addScrollUpEvent**
```
import { addScrollUpEvent } from './utils/events'

// Add scroll up listener (binded by default to window)
addScrollUpEvent()
```

**addScrollDownEvent**
```
import { addScrollDownEvent } from './utils/events'

// Add scroll down listener to specific DOM element
addScrollUpEvent(document.getElementById('#div'))
```

mcaskill commented on assets/scripts/utils/tickers.js:
```diff
@@ -0,0 +1,58 @@
+}
+
+
+/**
+ * Throttle function: fire the callback while the action is being performed for the defined iteration time
+ * @param {function}    callback    - callback function
+ * @param {number}      delay       - waiting time in milisecond
+ * @return {function}   callback
+ */
```
```suggestion
/**
 * Creates a throttled function.
 *
 * A throttled function invokes `callback` at most once per every
 * `delay` milliseconds.
 *
 * Useful for rate-limiting an event that occurs in quick succession.
 *
 * @template {function} T
 *
 * @param  {T}       callback - The function to throttle.
 * @param  {number}  delay    - The number of milliseconds to wait.
 * @return {function<T>} The new throttled function.
 */
```

mcaskill reviewed (changes requested):
The [`utils/events.js`](assets/scripts/utils/events.js) is a neat idea (similar to the recent changes I made to [`utils/visibility.js`](assets/scripts/utils/visibility.js)) but I recommend this file be moved to its own pull request (you can leave the new [`utils/tickers.js`](assets/scripts/utils/tickers.js) here).

I have a number of suggestions to make about it, such as:

* Rename references to element, document, window with `EventTarget` as that is the common interface.
* Improve block comments.
* Add the [examples you provided](https://github.com/locomotivemtl/locomotive-boilerplate/pull/120#issuecomment-1138423788) in this pull request's comments into the file itself.

mcaskill commented on assets/scripts/utils/is.js:
```diff
@@ -1,37 +1,22 @@
-const toString = Object.prototype.toString;
-const arrayLikePattern = /^\[object (?:Array|FileList)\]$/;
+/**
+ * Check if argument is an object
+ * @param {?object}     x   - elemet to check
+ * @return {boolean}    true if element is an object
+ */
```
```suggestion
/**
 * Determines if the argument is object-like.
 *
 * A value is object-like if it's not `null` and has a `typeof` result of "object".
 *
 * @param  {*} x - The value to be checked.
 * @return {boolean}
 */
```

mcaskill commented on assets/scripts/utils/is.js:
```diff
@@ -1,37 +1,67 @@
+ * @return {boolean}    true if element is a function
+ */
+
+const isFunction = x => x instanceof Function
```
@lucasvallenet 👋

Jerek0 commented on assets/scripts/utils/transform.js:
```diff
@@ -1,21 +1,51 @@
-    el.style.webkitTransform = transformValue;
-    el.style.msTransform = transformValue;
-    el.style.transform = transformValue;
+/**
+ * Transform function
+ * @param {HTMLElement} $el         - DOM element
+ * @param {number}      value       - transform value
+ * @return {void}
+ */
+
+const transform = ($el, value) => {
+    $el.style.msTransform = value
+    $el.style.webkitTransform = value
+    $el.style.MozTransform = value
+    $el.style.OTransform = value
+    $el.style.transform = value
 }
```
I think we can drop this one? Transform doesn't need prefixes for the modern browsers we agreed to support

arnvvd commented:
**Outcome from meeting on 2022-10-05**
**Status**: Approved by all.
- Next step: Locodash 
- Js file to declare constants.
